From 95e784bdfc4b438bb4bb5a2246e77a66def5e2cb Mon Sep 17 00:00:00 2001
From: Bhavuk Garg <bhavukgarg@google.com>
Date: Tue, 31 Mar 2026 12:00:00 +0000
Subject: [PATCH] Disable System Overlays to prevent tapjack attack

ContactsPickerActivity is not hiding overlay windows, allowing a malicious app with
SYSTEM_ALERT_WINDOW permission to hide privacy banner or show a custom message on top of
ContactsPicker page misleading user into giving access of contacts without full disclosure.
This change restricts overlay windows from being drawn on top of Contacts Picker to prevent
such attacks

Recording - http://b/488233632#attachment75003473

Bug: 488233632
Flag: EXEMPT CVE_FIX
Test: atest ContactsPickerIntegrationTests ContactsPickerComponentTests
Change-Id: Ia1a9d7f3ec3e924363c1617f5051f3fca7ded5a3
(cherry picked from commit 9c389023fc9bd5bb84ad44a027e2737d81340cd0)
---
 AndroidManifest.xml                               |  1 +
 .../contactspicker/ContactsPickerActivity.kt      |  2 ++
 .../contactspicker/ContactsPickerActivityTest.kt  | 15 +++++++++++++++
 3 files changed, 18 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 1696e0a..b87398c 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -48,5 +48,6 @@
     <uses-permission android:name="android.permission.QUERY_ALL_PACKAGES" />
     <uses-permission android:name="android.permission.MANAGE_CONTACTS_PICKER_SESSION" />
     <uses-permission android:name="android.permission.INTERACT_ACROSS_USERS_FULL" />
+    <uses-permission android:name="android.permission.HIDE_OVERLAY_WINDOWS" />
 
 </manifest>
\ No newline at end of file
diff --git a/src/com/android/contactspicker/ContactsPickerActivity.kt b/src/com/android/contactspicker/ContactsPickerActivity.kt
index b72467d..d9118cc 100644
--- a/src/com/android/contactspicker/ContactsPickerActivity.kt
+++ b/src/com/android/contactspicker/ContactsPickerActivity.kt
@@ -62,6 +62,8 @@ class ContactsPickerActivity : Hilt_ContactsPickerActivity() {
         Trace.beginSection("$TAG#coldStart")
         super.onCreate(savedInstanceState)
 
+        window.setHideOverlayWindows(true)
+
         if (savedInstanceState != null) {
             setupComposeUi()
             return
diff --git a/tests/componenttests/src/com/android/contactspicker/ContactsPickerActivityTest.kt b/tests/componenttests/src/com/android/contactspicker/ContactsPickerActivityTest.kt
index f5fafb1..f3fef87 100644
--- a/tests/componenttests/src/com/android/contactspicker/ContactsPickerActivityTest.kt
+++ b/tests/componenttests/src/com/android/contactspicker/ContactsPickerActivityTest.kt
@@ -33,6 +33,7 @@ import android.platform.test.annotations.RequiresFlagsEnabled
 import android.platform.test.flag.junit.CheckFlagsRule
 import android.platform.test.flag.junit.DeviceFlagsValueProvider
 import android.provider.ContactsContract
+import android.view.WindowManager
 import androidx.compose.ui.test.assertIsDisplayed
 import androidx.compose.ui.test.junit4.v2.createEmptyComposeRule
 import androidx.compose.ui.test.onNodeWithContentDescription
@@ -178,6 +179,20 @@ class ContactsPickerActivityTest {
         Intents.release()
     }
 
+    @Test
+    @RequiresFlagsEnabled(Flags.FLAG_ENABLE_SYSTEM_CONTACTS_PICKER)
+    fun onCreate_setsHideOverlayWindows() {
+        val scenario = ActivityScenario.launch<ContactsPickerActivity>(baseIntent)
+        scenario.onActivity { activity ->
+            val attrs = activity.window.attributes
+            assertThat(
+                    attrs.privateFlags and
+                        WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS
+                )
+                .isNotEqualTo(0)
+        }
+    }
+
     @Test
     @RequiresFlagsDisabled(Flags.FLAG_ENABLE_SYSTEM_CONTACTS_PICKER)
     fun intent_contactPickerFlagDisabled_throws() {
-- 
2.55.0

